app/src/main/java/com/github/weg_li_android/MainActivity.kt

```diff
@@ -66,4 +265,163 @@ class MainActivity : AppCompatActivity() {
+                        for (i in 0 until count) {
+                            val imageUri: Uri? = data.clipData!!.getItemAt(i).uri
+                            val insertIndex = imageUri?.let { mainViewModel.addViolationPhoto(it) }
+                            if (insertIndex != null) {
+                                photoAdapter.notifyItemInserted(insertIndex)
+                            }
+                        }
+                    } else if (data.data != null) {
+                        val mImageUri: Uri? = data.data
+                        val insertIndex = mImageUri?.let { mainViewModel.addViolationPhoto(it) }
+                        if (insertIndex != null) {
+                            photoAdapter.notifyItemInserted(insertIndex)
+                        }
+                    }
+                }
+                takeImage -> {
+                    val bitmap = data.extras?.get("data") as Bitmap
+                    if (isStoragePermissionGranted()) {
+                        Timber.e("granted")
+                        saveAndInsertPhoto(bitmap)
+                    }
+                    else {
+                        tempBitmapOnRequestPermissions = bitmap
+                        Timber.e("Not granted")
+                    }
+                }
+            }
+        }
+    }
+
+    private fun saveAndInsertPhoto(bitmap : Bitmap) {
+        val imageUri : Uri = saveImageBitmap(bitmap, System.currentTimeMillis().toString())
+        Timber.e(imageUri.toString())
+        val insertIndex = mainViewModel.addViolationPhoto(imageUri)
+        photoAdapter.notifyItemInserted(insertIndex)
+    }
+
+    private fun isStoragePermissionGranted(): Boolean {
+        return if (Build.VERSION.SDK_INT >= 23) {
+            if (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
+                == PackageManager.PERMISSION_GRANTED
+            ) {
+                Timber.v("Permission is granted")
+                true
+            } else {
+                Timber.v( "Permission is revoked")
+                ActivityCompat.requestPermissions(
+                    this,
+                    arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
+                    permissionsRequestCode
+                )
+                (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
+                        == PackageManager.PERMISSION_GRANTED)
+            }
+        } else { //permission is automatically granted on sdk<23 upon installation
+            Timber.v( "Permission is granted")
+            true
+        }
+    }
+
+    private fun saveImageBitmap(image_bitmap: Bitmap, image_name: String) : Uri {
+        val fos: OutputStream
+        val resolver: ContentResolver = contentResolver
+        val contentValues = ContentValues().apply {
+            put(MediaStore.MediaColumns.DISPLAY_NAME, image_name)
+            put(MediaStore.MediaColumns.MIME_TYPE, "image/jpeg")
```

Comment:
Please look up whether there is a constant for this MIME type and use it if possible.